precompile/contracts/feemanager/simulated_test.go

```diff
@@ -0,0 +1,314 @@
+}
+
+func newBackendWithFeeManager(t *testing.T) *sim.Backend {
+	t.Helper()
+	chainCfg := params.Copy(params.TestChainConfig)
+
+	// Enable FeeManager at genesis with admin set to adminAddress and initial fee config.
+	params.GetExtra(&chainCfg).GenesisPrecompiles = extras.Precompiles{
+		feemanager.ConfigKey: feemanager.NewConfig(utils.NewUint64(0), []common.Address{adminAddress}, nil, nil, &genesisFeeConfig),
+	}
+
+	return sim.NewBackend(
+		types.GenesisAlloc{
+			adminAddress:        {Balance: big.NewInt(1000000000000000000)},
+			unprivilegedAddress: {Balance: big.NewInt(1000000000000000000)},
+		},
+		sim.WithChainConfig(&chainCfg),
+	)
+}
+
+// Helper functions to reduce test boilerplate
+
+func deployFeeManagerTest(t *testing.T, b *sim.Backend, auth *bind.TransactOpts) (common.Address, *feemanagerbindings.FeeManagerTest) {
+	t.Helper()
+	addr, tx, contract, err := feemanagerbindings.DeployFeeManagerTest(auth, b.Client(), feemanager.ContractAddress)
+	require.NoError(t, err)
+	testutils.WaitReceiptSuccessful(t, b, tx)
+	return addr, contract
+}
+
+// setFeeConfig sets fee config using a FeeConfig struct instead of individual parameters
+func setFeeConfig(t *testing.T, b *sim.Backend, contract *feemanagerbindings.FeeManagerTest, auth *bind.TransactOpts, config commontype.FeeConfig) {
+	t.Helper()
+	tx, err := contract.SetFeeConfig(
+		auth,
+		config.GasLimit,
+		big.NewInt(int64(config.TargetBlockRate)),
+		config.MinBaseFee,
+		config.TargetGas,
+		config.BaseFeeChangeDenominator,
+		config.MinBlockGasCost,
+		config.MaxBlockGasCost,
+		config.BlockGasCostStep,
+	)
+	require.NoError(t, err)
+	testutils.WaitReceiptSuccessful(t, b, tx)
+}
+
+// verifyFeeConfigMatches verifies that the contract's fee config matches the expected config
+func verifyFeeConfigMatches(t *testing.T, contract *feemanagerbindings.FeeManagerTest, expected commontype.FeeConfig) {
+	t.Helper()
+	actual, err := contract.GetFeeConfig(nil)
+	require.NoError(t, err)
+
+	require.Zero(t, expected.GasLimit.Cmp(actual.GasLimit), "gasLimit mismatch")
+	require.Equal(t, big.NewInt(int64(expected.TargetBlockRate)), actual.TargetBlockRate, "targetBlockRate mismatch")
+	require.Zero(t, expected.MinBaseFee.Cmp(actual.MinBaseFee), "minBaseFee mismatch")
+	require.Zero(t, expected.TargetGas.Cmp(actual.TargetGas), "targetGas mismatch")
+	require.Zero(t, expected.BaseFeeChangeDenominator.Cmp(actual.BaseFeeChangeDenominator), "baseFeeChangeDenominator mismatch")
+	require.Zero(t, expected.MinBlockGasCost.Cmp(actual.MinBlockGasCost), "minBlockGasCost mismatch")
+	require.Zero(t, expected.MaxBlockGasCost.Cmp(actual.MaxBlockGasCost), "maxBlockGasCost mismatch")
+	require.Zero(t, expected.BlockGasCostStep.Cmp(actual.BlockGasCostStep), "blockGasCostStep mismatch")
+}
+
+// verifyEventFeeConfigMatches verifies that an event's fee config matches the expected config
+func verifyEventFeeConfigMatches(t *testing.T, actual feemanagerbindings.IFeeManagerFeeConfig, expected commontype.FeeConfig) {
```

Comment:
why do we have 2 very similar functions that does require for all fields two times? Can we just have a single `verifyFeeConfigsMatch` function that takes two `commontype.FeeConfig` and then just convert the struct/event type to FeeConfig in another helper function.